services/src/pro/quota/mod.rs

```diff
@@ -0,0 +1,161 @@
+use futures::{
+    channel::mpsc::{channel, Receiver, Sender},
+    StreamExt,
+};
+use geoengine_operators::pro::meta::quota::{ComputationContext, ComputationUnit, QuotaTracking};
+use std::sync::Arc;
+
+use geoengine_datatypes::util::test::TestDefault;
+
+use crate::pro::users::UserId;
+
+use super::users::{UserDb, UserSession};
+
+// TODO: find a good default, choose w.r.t. resources or make configurable
+const QUOTA_CHANNEL_SIZE: usize = 1000;
+
+#[derive(Debug, Clone)]
+pub struct QuotaTrackingFactory {
+    quota_sender: Sender<ComputationUnit>,
+}
+
+impl QuotaTrackingFactory {
+    pub fn new(quota_sender: Sender<ComputationUnit>) -> Self {
+        Self { quota_sender }
+    }
+
+    pub fn create_quota_tracking(
+        &self,
+        session: &UserSession,
+        context: ComputationContext,
+    ) -> QuotaTracking {
+        QuotaTracking::new(
+            self.quota_sender.clone(),
+            ComputationUnit {
+                issuer: session.user.id.0,
+                context,
+            },
+        )
+    }
+}
+
+impl TestDefault for QuotaTrackingFactory {
+    fn test_default() -> Self {
+        let (quota_sender, _quota_receiver) = channel(QUOTA_CHANNEL_SIZE);
+        Self::new(quota_sender)
+    }
+}
+
+pub struct QuotaManager<U: UserDb + 'static> {
+    user_db: Arc<U>,
+    quota_receiver: Receiver<ComputationUnit>,
+}
+
+impl<U: UserDb + 'static> QuotaManager<U> {
+    pub fn new(user_db: Arc<U>, quota_receiver: Receiver<ComputationUnit>) -> Self {
+        Self {
+            user_db,
+            quota_receiver,
+        }
+    }
+
+    pub fn run(mut self) {
+        crate::util::spawn(async move {
+            while let Some(computation) = self.quota_receiver.next().await {
+                // TODO: issue a tracing event instead?
+                // TODO: also log workflow, or connect the computation context to a workflow beforehand.
+                //       However, currently it is possible to reuse a context for multiple queries.
+                //       Also: the operators crate knows nothing about workflows as of yet.
+                log::info!(
+                    "Quota received. User: {}, Context: {}",
+                    computation.issuer,
+                    computation.context
+                );
+
+                let user = UserId(computation.issuer);
+                // TODO: what to do if this fails (quota can't be recorded)? Try again later?
+                let r = self.user_db.increment_quota_used(&user, 1).await;
```

Comment:
It would be nice to batch this I guess. Maybe we can fetch *n* messages at once?

Reply:
we would need to group the stuff by users though. I can add a TODO for batching, but I think it's pretty dangerous because this stuff will get lost if the application crashes or exits.

Reply:
> I guess we need to measure some concurrent workflows, because there is a blocking if this is not fast enough.

it can't ever block because quota management and workflow processing is independent of each other. The query processor does not wait for the quota to be recorderd.